lib/server.js

```diff
@@ -312,20 +648,637 @@ function createServer(options, clients) {
         var throttleHandle = throttle.createThrottle(options.throttle);
         server.use(throttle.throttleHandler(throttleHandle));
     }
+    // Add debug logging before authentication
+    server.use(function preAuthDebug(req, res, next) {
+        if (req.isS3Request) {
+            req.log.debug('S3_DEBUG: About to run authentication handler');
+        }
+        next();
+    });
+
+    // Anonymous access handler for public buckets (before authentication,
+    // with metadata client)
+    server.use(function (req, res, next) {
+        // Set up metadataPlacement early for anonymous access handler
+        req.metadataPlacement = clients.metadataPlacement;
+        anonymousAuth.anonymousAccessHandler(req, res, next);
+    });
+
     server.use(auth.authenticationHandler({
         log: log,
         mahi: clients.mahi,
         keyapi: clients.keyapi
     }));
 
+    // Add debug logging after authentication
+    server.use(function postAuthDebug(req, res, next) {
+        if (req.isS3Request) {
+            req.log.debug({
+                caller: req.caller ? {
+                    type: req.caller.account ? 'account' : 'user',
+                    login: req.caller.account ?
+                    req.caller.account.login : 'unknown',
+                    uuid: req.caller.account ?
+                    req.caller.account.uuid : 'unknown'
+                } : null,
+                hasAuthorization: !!req.authorization
+            }, 'S3_DEBUG: Authentication completed');
+        }
+        next();
+    });
+
     server.use(auth.gatherContext);
 
-    // Add various fields to the 'req' object before the handlers get called.
+    // Add various fields to the 'req' object before ANY
+    // S3 handlers get called.
+    // This MUST come before both root handler
+    // and S3 routing so req.metadataPlacement is available
     server.use(common.setupHandler(options, clients));
 
-    // Buckets API
+    // Add consolidated root handler AFTER authentication AND setup
```

Comment:
This was what I was talking about earlier around the `NOTE:` comment.  Am I right?